### ISSUE TITLE
[Feature] More configurable cartridge support than simply ROM

### ISSUE THREAD

**Issue by brawaga** [openMSX/openMSX#1307]
How I can see now in docs, we can only insert ROMs as a cartridge. And as far as I understand, for 16k cartridge it is mirrored among all 64k space.
What I suggest, to introduce an XML as it is for machine set up, with each 8k/16k (regarding mapper chosen) set to be ROM (maybe mirrored), off, or RAM in the slot/subslot cartridge inserted in.
Let's say i have ROM in 4000h…7FFFh and some pages of RAM in 8000h…BFFFh. This should be possible regarding the feature I suggest.
As additional option, let's say the RAM could be enforced with backup power supply (so, some storage must be organized for openMSX on hard drive).

**Comment by m9710797** [comment]
Hi brawaga,

Thanks for your feedback. Though I do have some questions, because as far as I currently understand your request, most of what you ask is already possible.

When you insert a ROM as a cartridge, openMSX tries to detect/guess the mapper type. For a 16k cartridge the default guess is indeed a mirrored mapper. But you can also explicitly specify the type with the '-romtype' option. For example:

    openmsx myrom.rom -romtype normal4000

Assuming 'myrom.rom' is a 16k file, this ROM will only be visible in address range 0x4000-0x7fff (so not mirrored over the full 64k as is the default). To see the list of possible types execute the command 'openmsx_info romtype' in the openMSX console.

You suggest to create a XML file with detailed description of the cartridge. Exactly that is already possible. Except that we call such things 'extensions' rather than 'cartridges' (different name because the 'extension' mechanism is more general than only ROM-cartridges). See the '<openmsx>/share/extension' directory for lots of examples.

So by creating a specific xml file (copy/paste from the examples) you can configure ROM/RAM in exactly the required memory regions. Though this xml file needs to specify one specific filename. The advantage of a 'cartridge' over an 'extension' is that the same "cartridge configuration" can be reused for multiple rom files. At the moment openMSX does not support 'extensions' with a "parameterized" filename.

But actually ... cartridge-extensions with a parametrized filename is what we call "mapper types" (see the '-romtype' option above). Though the list of mapper types is built-in in openmsx (so cannot be extended by xml files). Part of the reason for this is that many mapper types have special features (e.g. bank-switching) that cannot be described in xml alone (so these need special c++ code anyway).

As far as we're aware, openMSX already supports most (all?) common mapper types. And for very specific cartridge configurations it's possible to define a specific extension (and because it's specific it doesn't need to have a parametrized filename).

Several mapper-types already support battery-backed RAM (the content of those RAMs is stored in the '<openmsx>/persistent' directory.

@brawaga: With the above information in mind, can you elaborate on your request? Why exactly do you need this feature? Isn't it already covered by the existing functionality? Do you think openMSX lacks support for an important mapper-type?

Thanks.

Wouter

**Comment by brawaga** [comment]
Thank you very much, Wouter, for your detailed answer.

I am doing a cartridge to change the language of my MSX to Russian, including font, keyboard, and optionally messages. I am too bad in soldering to replace BIOS, which is what I started developing from, so changed my mind to cartridge. Now it is configured via pressing keys on boot, maybe I will want to make a configurator to set it up via some UI and store in battered RAM. Also, I ran out of features that can be implemented without RAM, so heterogeneous memory needed.
Regarding what you said, it is pretty obvious all features I requested are already there, what I have to do is use another terminology for my project: «extension» term instead of «cartridge», read the manual for extensions, and ensure I can combine all the needed features into one thing.

The manual I found is User Manual, which does not contain any information about mastering extensions. And quick googling gave nothing. Do you have such a manual by the way? I explored many examples, looks like I need to put a structure like `<primary…><MyExtension><sramname…/></MyExtension></primary>`, but I met also several ROMs mentioned, or ROM+SRAM for same `<mem base=…>`, which is confusing, but what I can suggest there regarding Tecall_MSX_RAMFILE_TM220 is they are stacked upon their file sizes. Based on these assumptions I can master some extension.xml, but those are assumptions, so I can miss something without the docs.

Maybe you also can give some information how external devices support can be implemented? I found `<io base=…>` thing, but it is only port number, not the device behaviour. I have some TODO about external I/O, but in the end of the list, so did not mentioned it in the Feature Request.

**Comment by m9710797** [comment]
Hi brawaga,

Indeed the user manual doesn't explain this stuff. That manual is meant for more 'typical' users. What you're trying to do is already highly specialized.

The '[openMSX setup guide](https://openmsx.org/manual/setup.html#hardwareedit)' has a (short) section on editing xml files. But this is pretty general information.So I'm afraid the best you can do is study the examples and ask questions (via github or via our IRC channel).

Note that the the XML descriptions only allow you to combine existing functionality. Some but not all of this functionality is configurable. For example (most) devices allow you to change the number of the IO port(s) they react to, but the actual functionality that's trigger via those ports is fixed (coded in c++ rather than xml).

I'm closing this ticket, because (i think) your original question has been answered, right? But feel free to continue asking more questions, even via this ticket (I'll keep it open for extra comments). Or contact us via some other way.

Wouter

**Comment by brawaga** [comment]
Yes, Wouter, fine to close.
Unfortunately just on adding feature to test SRAM content persists across reboots Gen80 fails to build it and I cannot link splitted sources with L80 yet, can you say if this config
```
<?xml` version="1.0" ?>
<!DOCTYPE msxconfig SYSTEM 'msxconfig2.dtd'>
<msxconfig>
  <info>
    <name>Rusification</name>
    <code/>
    <release_year>2021?</release_year>
    <description>ROM+SRAM</description>
    <type>Cartridge</type>
  </info>
  <devices>
    <primary slot="any">
        <ROM id="Rusification">
          <mem base="0x4000" size="0x8000"/>
          <rom>
            <filename>rusifctn.rom</filename>
          </rom>
          <sramsize>16</sramsize>
          <sramname>rusifctn.sram</sramname>
        </ROM>
    </primary>
  </devices>
</msxconfig>
```
I built yesterday will store 16k 0x8000…0xBFFF of my slot somewhere across reboots?
I intentionally didn't mention subslot in the XML, hence I do not support this to be in subslot.

**Comment by MBilderbeek** [comment]
The ROM device does not have SRAM, so it's no use to specify the `sramsize`/`sramname`.

**Comment by m9710797** [comment]
Hi brawaga,

The default mapper type is 'mirrored'. So it's as if you had a tag <mappertype>mirrored</mappertype> inside your <ROM>.
This type contains no SRAM, and thus the <sramname> tag you specified has no effect.

You can look here for some tecnical details on mapper types:
   http://bifi.msxnet.org/msxnet/tech/megaroms
OpenMSX supports many more mapper types, but this page lists the most common ones.

In openMSX we tend to only implement hardware (thus also mapper types) that does exist in the real world. As far as I know there is no real MSX hardware device that acts as a fully battery-backed RAM.

So for your project you could try to choose one of the existing mappers with SRAM support. Alternatively you could emulate a mapper with flashrom (re-flashing is fully emulated, and the changed ROM content is stored on disk by openmsx). Another alternative is to extend openMSX with a new device that does act like batter-backed RAM. This requires writing new c++ code. (But as I said before, in general, we prefer to only add support for real existing hardware).

Wouter

**Comment by brawaga** [comment]
So how can I have SRAM?

**Comment by MBilderbeek** [comment]
As Wouter wrote, there are 3 options:

1. choose one of the existing mappers with SRAM support
2. emulate a mapper with flashrom (re-flashing is fully emulated, and the changed ROM content is stored on disk by openmsx). 
3. extend openMSX with a new device that does act like battery-backed RAM. This requires writing new C++ code.

**Comment by m9710797** [comment]
Hi brawaga,

In order to help you better, I need to understand better exactly what you're trying to do. So I'll ask a few questions, and depending on the answers there are a few possible scenarios:

a) Does the configuration you want to emulate have to exist as real MSX hardware?

In openMSX we prefer to (only) emulate hardware that actually exists. (Or allow to emulate 'fantasy' hardware when it doesn't require too much extra effort from our side). As far as I know there is no real MSX cartridge that only contains battery-backed RAM. Real MSX cartridges always combine battery backed RAM with some other ROM mapper mechanism.

So in openMSX we emulate those existing mappers, but there's currently no direct support for (only) emulating battery-backed RAM. If you want exactly this then you'll have to extend openMSX itself (write extra C++ code). But maybe you're use case is fine with a different hardware configuration which does exist as a real MSX cartridge? For example the following mapper contains 8kB of SRAM. It can be selected via the "ASCII8SRAM8" mapper type. See here for technical info:
   http://bifi.msxnet.org/msxnet/tech/megaroms#ascii8sram

b) Assuming you want an existing MSX cartridge. Does it really have to be battery-backed RAM? Or are you fine with alternative solution for persistent storage?

If you really need persistent RAM then you'll have to pick one of the ROM mapper types that contain some amount of SRAM (e.g. the ASCII8SRAM8 example I gave above). But an alternative is to use a ROM mapper based on flash ROM. Flash ROM can be re-programmed by the MSX itself, and (like in real hardware) the changed ROM content is preserved (and this is emulated in openMSX). Though a disadvantage of flash ROM is that changing it is not as easy as simply writing new values to it (as you do for RAM). A good example of a ROM mapper built using flash ROM is "MegaFlashRomScc". This contains 512kB of fully programmable flash ROM (and also a SCC and a Konami-like bank-switching mechanism).


So in summary: there is currently no mechanism to directly use SRAM in an arbitrary way in openMSX (it's always coupled to a specific ROM mapper implementation). But depending on what exactly you want to do, there are some alternatives.

**Comment by brawaga** [comment]
Hi.
a) Yes, why cannot cartridge have both ROM and RAM aboard, this is even adviced as «safest and most reliable method» (you can use it as search string as it was cited) at https://www.msx.org/wiki/Develop_a_program_in_cartridge_ROM. The only difference I want it backed up with some battery power, which is also far from to be a crime =)
Further reading shows me you don't take kindly battery backed RAM only cartridges, and seemingly confirm ROM + backed RAM combination, which is strange, hence the RAM-only is just a particular case of latter, but anyway I want exactly latter, and «strangeness» is just a term of hypotetical discussion not to be developed here.
Configuration you offer is great step forward towards what I want to get. BTW, whilst I can  consider 8kB backed RAM enough for my needs, for ROM part this is already insufficient. I have 8kB ROM data only, and some code of 2-4 kB (I did not take much attention to its size, so rough estimation), and I want some reserves to grow it.
b) Actually, I do not want to have so much as 8kB RAM backed, most of it will be fine unbacked, and some part of it can be replaced with Flash, but this separation just sophisticate the configuration, so I started the discussion from simple enough configuration of ROM + backed SRAM. The Game Master 2 (Konami RC 755) and probably Hydlide 2 are good enough for capacity, I am just not sure if SRAM is backed.

**Comment by m9710797** [comment]
Hi,

Your request is perfectly valid. But I cannot make openMSX do more than what is currently implemented. Hence my suggestion for some alternative solutions.

Note that the 'ASCII8SRAM8' mapper can have a lot more than 8KB ROM. The  'ASCII8' part means the ROM is switched in banks of 8kB according to the "Ascii8" mechanism. But there can be up to 128 banks (so max 1024kB ROM in total). However there is only 1 SRAM bank, so only 8kB SRAM (for this particular mapper type).

"Game master 2" and "Hydlide 2" are good alternatives for "ASCII8SRAM8". They also contain SRAM, but they use a different bank-switching mechanism and have different limits for the max amount of ROM, and the amount of SRAM.

For details see:
   http://bifi.msxnet.org/msxnet/tech/megaroms
   https://www.msx.org/wiki/MegaROM_Mappers

**Comment by brawaga** [comment]
That's exactly where I found them to mention in my previous message. How can I try them? Or maybe I should add ROM and SRAM as sibling nodes inside <primary> node? This will help keep linear model without any mapper.
And also, @MBilderbeek, you say there's no SRAM in ROM, but why it is mentioned there https://github.com/openMSX/openMSX/blob/master/share/extensions/MegaFlashROM_SCC%2B.xml, https://github.com/openMSX/openMSX/blob/master/share/extensions/MegaFlashROM_SCC.xml, https://github.com/openMSX/openMSX/blob/master/share/extensions/Tecall_MSX_RAMFILE_TM220.xml  then?

**Comment by MBilderbeek** [comment]
When you start a ROM in openMSX, you can optionally specify a mapper type. So, if you need to select one of these. If you want to make an extension type in XML, you can also specify the mappertype in the XML.

The configs you mention all have a specific mappertype tag inside the ROM tag. This mappertype tag specifies which exact mapper type will be emulated and some of these have SRAM and those will read the sramname tag or other specific tags for that mapper type.

Please let me know if I'm clear enough by asking more questions if necessary :)
